06_puppy/lexual/puppy.go

```diff
@@ -0,0 +1,96 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"sync"
+)
+
+var out io.Writer = os.Stdout
+
+type Puppy struct {
+	ID     uint32
+	Breed  string
+	Colour string
+	Value  int
+}
+
+// Storer defines standard CRUD operations for Puppys
+type Storer interface {
+	CreatePuppy(*Puppy)
+	ReadPuppy(id uint32) *Puppy
+	UpdatePuppy(id uint32, puppy *Puppy)
+	DeletePuppy(id uint32) bool
+}
+
+// MapStore ////
+
+type MapStore map[uint32]Puppy
+
+// CreatePuppy stores a copy of Puppy p
+func (store *MapStore) CreatePuppy(p *Puppy) {
+	(*store)[p.ID] = Puppy{p.ID, p.Breed, p.Colour, p.Value}
+}
+
+// ReadPuppy retrieves a previously stored Puppy
+func (store *MapStore) ReadPuppy(id uint32) *Puppy {
+	p, ok := (*store)[id]
+	if !ok {
+		return nil
+	}
+	return &p
+}
+
+// UpdatePuppy updates details of Puppy in store
+func (store *MapStore) UpdatePuppy(id uint32, puppy *Puppy) {
+	(*store)[id] = Puppy{puppy.ID, puppy.Breed, puppy.Colour, puppy.Value}
```

Comment:
You probably need to check if id==puppy.ID first, otherwise, they could end up inconsistent